### PR TITLE
[installer]: add HTTP_PROXY envvars to the Installer

### LINKED ISSUE #12820: Allow to add the  environment variable

### LINKED ISSUE #12821: Add automated tests with HTTPS proxy

### LINKED ISSUE #12835: Make sure that all pods are considering the `https_proxy` environment variable
Duplicate of #10484

### PR DESCRIPTION
**Pull request by MrSimonEmms**
## Description

Adds preliminary support for `HTTP_PROXY` (and `HTTPS_PROXY` and `NO_PROXY`) in the Installer. This adds an optional secret which receives the proxy settings from the KOTS CLI settings - the `NO_PROXY` envvar also hard-codes some additional parameters required for the application to work.

**Important** - this is not a full feature. It is merely the initial work to enable other teams to help out in their areas of expertise. This adds the appropriate envvars to the resources and wires them all up in a fashion that allows the Gitpod stack to run without error. However, it **DOES NOT** allow an image to be built or workspace to be run. There are additional tickets in #10769 that detail these tasks

An interesting thing of note is that the `@grpc/grpc-node` library does not respect wildcard `NO_PROXY` URLs. To address that, the `server` adds an addition `no_grpc_proxy` (which is the primary envvar for this library) and changes all resources it calls to the FQDN. See [the issue](https://github.com/grpc/grpc-node/issues/1293) for more details.

## Related Issue(s)

Fixes #12835 
Fixes #12820 
Fixes #12821

## How to test

> This will also need testing in a non-proxied instance to ensure no regressions are introduced

- Create a new server with [Squid](https://www.digitalocean.com/community/tutorials/how-to-set-up-squid-proxy-on-ubuntu-20-04) installed. This server must have unrestricted internet access
- Configure an airgapped installation and configure the k3s cluster to access the proxy
- Run `kubectl kots install gitpod --http-proxy <http-proxy> --https-proxy <https-proxy> --no-proxy <no-proxy>` to install Gitpod

### Configuring your cluster

A [repo](https://github.com/MrSimonEmms/gitpod-self-hosted-infrastructure) exists that will create a k3s cluster for you in Azure. Run `make azure-k3s` with the following settings in your `.auto.tfvars` file

```terraform
domain_name = "<your domain name>"
azure_k3s = true
enable_airgapped = true
http_proxy = "<your proxy settings - eg http://user:password@ip-address:3128">
```

## Release Notes

```release-note
[installer]: add HTTP_PROXY envvars to the Installer
```

## Werft options:

- [x] /werft with-preview

### CODE REVIEW

**Comment by roboquat** [comment]
@MrSimonEmms: Adding the "do-not-merge/release-note-label-needed" label because no release-note block was detected, please follow our [release note process](https://git.k8s.io/community/contributors/guide/release-notes.md) to remove it.

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository.
</details>

**Comment by corneliusludmann** [comment]
> Fixes #10484
> Fixes #12820
> Fixes #12821

Is this in the description still true that all the issues will be fixed with this PR or is it a leftover of the original PR?

File: install/installer/pkg/components/content-service/configmap.go [hunk 1]
```diff
@@ -29,7 +29,7 @@ func configmap(ctx *common.RenderContext) ([]runtime.Object, error) {
 
 	cscfg := config.ServiceConfig{
 		Service: baseserver.ServerConfiguration{
-			Address: fmt.Sprintf(":%d", RPCPort),
+			Address: fmt.Sprintf("0.0.0.0:%d", RPCPort),
```
Comment (sagor999) [inline install/installer/pkg/components/content-service/configmap.go:32]: will this break anything? :) 
I kind of remember a similar PR that changed from `:` to either `0.0.0.0` or `127.0.0.1` and it broke things. 
So just want to double check here.

Comment (MrSimonEmms) [reply install/installer/pkg/components/content-service/configmap.go:32]: From my tests, seemingly not. I previously tried `127.0.0.1` and that _DID_ break it. It looks like this is the listening address - if you have `127.0.0.1:<port>` then it only allows for things going to that address - if you use `0.0.0.0` then it accepts requests via anything (eg, `localhost`, `127.0.0.1`, `<IP`> and `<FQDN>`)

**Review by sagor999 — APPROVED** [review]

File: install/kots/manifests/gitpod-http-proxy-settings.yaml [hunk 2]
```diff
@@ -0,0 +1,15 @@
...
+kind: Secret
+metadata:
+  name: http-proxy-settings
+  labels:
+    app: gitpod
+    component: gitpod-installer
+type: Opaque
+data:
+  httpProxy: repl{{ HTTPProxy | Base64Encode | quote }}
+  httpsProxy: repl{{ HTTPProxy | Base64Encode | quote }}
```
Comment (Pothulapati) [inline install/kots/manifests/gitpod-http-proxy-settings.yaml:14]: Shouldn't this be `HTTPSPROXY`? 🤔

Comment (MrSimonEmms) [reply install/kots/manifests/gitpod-http-proxy-settings.yaml:14]: Well spotted

**Review by Pothulapati — APPROVED** [review]
:shipit:

File: install/installer/pkg/components/server/deployment.go [hunk 3]
```diff
@@ -119,6 +122,17 @@ func deployment(ctx *common.RenderContext) ([]runtime.Object, error) {
 				Name:  "WSMAN_CFG_MANAGERS",
 				Value: wsmanCfgManager,
 			},
+			{
+				Name: "no_grpc_proxy",
```
Comment (geropl) [inline install/installer/pkg/components/server/deployment.go:126]: @MrSimonEmms Shouldn't this be guarded by
```
if cfg.HTTPProxy == nil {
		return []corev1.EnvVar{}
	}
```
? :thinking:

**Comment by geropl** [comment]
Waiting for the build :point_up: to see if the default case is running.

File: install/installer/pkg/components/server/deployment.go (hunk 3 above)
Comment (MrSimonEmms) [reply install/installer/pkg/components/server/deployment.go:126]: It's not strictly necessary as it'll still generate successfully (check the `cmd/testdata/render/minimal/output.golden` file as an example). But it is tidier not to have it in there so I'll change it

File: install/installer/pkg/components/server/deployment.go [hunk 4]
```diff
@@ -122,6 +125,20 @@ func deployment(ctx *common.RenderContext) ([]runtime.Object, error) {
 		},
 	)
 
+	if ctx.Config.HTTPProxy != nil {
```
Comment (MrSimonEmms) [inline install/installer/pkg/components/server/deployment.go:128]: @geropl have updated as-per your comment 👍🏻

**Review by geropl — APPROVED** [review]
(WebApp related) Code LGTM, tested the preview and workspaces still start :upside_down_face: :heavy_check_mark:

**Comment by MrSimonEmms** [comment]
Thanks @geropl

**Pull request merged**